sonarts-core/src/symbols/lva.ts

```diff
@@ -19,22 +19,36 @@
  */
 import * as ts from "typescript";
 import { CfgBlock, CfgBlockWithPredecessors, ControlFlowGraph } from "../cfg/cfg";
-import { collectLeftHandIdentifiers, firstLocalAncestor, FUNCTION_LIKE, is, isAssignment } from "../utils/navigation";
+import {
+  collectLeftHandIdentifiers,
+  firstLocalAncestor,
+  FUNCTION_LIKE,
+  is,
+  isAssignment,
+  firstAncestor,
+} from "../utils/navigation";
 import { SymbolTable, Usage, UsageFlag } from "./table";
 
 export class LiveVariableAnalyzer {
   private blockAvailableReads: Map<CfgBlock, Set<ts.Symbol>>;
-  private root: ts.FunctionLikeDeclaration;
+  private root: ts.FunctionLikeDeclaration | ts.SourceFile;
+  private static readonly FUNCTION_OR_SOURCE_FILE = [...FUNCTION_LIKE, ts.SyntaxKind.SourceFile];
 
   constructor(private readonly symbols: SymbolTable) {}
 
-  public analyze(root: ts.FunctionLikeDeclaration) {
+  public analyzeFunction(root: ts.FunctionLikeDeclaration): LVAReturn | undefined {
     if (!is(root.body, ts.SyntaxKind.Block)) {
       return;
     }
-    const cfg = ControlFlowGraph.fromStatements((root.body as ts.Block).statements);
+    return this.analyze(root, (root.body as ts.Block).statements);
+  }
+
+  public analyze(root: ts.Node, statements: ts.Statement[]): LVAReturn | undefined {
+    const cfg = ControlFlowGraph.fromStatements(statements);
```

Comment:
Small thing : I think this is a good opportunity to isolate `analyze` from the list of statements and thus build the CFG outside of it, in this case, in `analyzeFunction`.